### ISSUE TITLE
Pattern matching and type narrowing on a generic class instance does not seem to work

### ISSUE THREAD

**Issue by squeaky-pl** [microsoft/pyright#7225]
**Describe the bug**
The problem involves pattern matching and a generic class instance. My intuition tells me that pyright should be able to narrow down the type here, but I maybe wrong.

**Code or Screenshots**

Consider following code

```python
import dataclasses
from typing import Generic, Literal, TypeVar

IdentifierT = TypeVar("IdentifierT")


@dataclasses.dataclass
class Token(Generic[IdentifierT]):
    identifier: IdentifierT


def process(token: Token[Literal["text"]]) -> None:
    pass


def main(token: Token[Literal["text", "number"]]) -> None:
    match token:
        case Token("text"):
            process(token)
        case _:
            pass
```

I am getting
```
error: Argument of type "Token[Literal['text', 'number']]" cannot be assigned to parameter "token" of type "Token[Literal['text']]" in function "process"
    "Token[Literal['text', 'number']]" is incompatible with "Token[Literal['text']]"
      Type parameter "IdentifierT@Token" is invariant, but "Literal['text', 'number']" is not the same as "Literal['text']" (reportArgumentType)
1 error, 0 warnings, 0 informations
```

While I would think that `case Token("text")` would ensure that it's `Token[Literal["text"]]`.

**Comment by squeaky-pl** [comment]
I figured out that this error goes away if I use

`IdentifierT = TypeVar("IdentifierT", contravariant=True)`

I am gonna close this.

**Comment by erictraut** [comment]
I'd like to take a closer look at this before we close it.

**Comment by erictraut** [comment]
After looking at this more closely, I think pyright's doing the right thing here. The type parameter for `Token` is invariant, so it's not possible to narrow the type of `token` from `Token[Literal["text", "number"]]` to `Token[Literal["text"]]`. The latter is not a subtype of the former.

As you discovered, if you make the type variable contravariant, it will work. However, I'll point out that your code is no longer type safe if you do this because the `Token` class uses `IdentifierT` in a manner that requires it to be invariant.

**Comment by squeaky-pl** [comment]
Thanks for spending time on this. I really appreciate your elaborate answers, and I learnt something new myself.
